lib/cc/analyzer/engine_process.rb

```diff
@@ -0,0 +1,101 @@
+require "posix/spawn"
+
+module CC
+  module Analyzer
+    class EngineProcess
+      # As per the SPEC, an engine itself has 15 minutes to run.
+      ENGINE_TIMEOUT = 15 * 60
+
+      # We allow an additional 5 for our own processing. Some engines choose to
+      # output everything at once at the end so this processing can take some
+      # time and cause builder processes to appear hung if we don't enforce a
+      # timeout for the overall operation as well as runnning the engine.
+      TIMEOUT = ENGINE_TIMEOUT + 5 * 60
+
+
+      def initialize(engine, stdout_io, stderr_io = StringIO.new)
+        @engine = engine
+        @stdout_io = stdout_io
+        @stderr_io = stderr_io
+
+        EngineTimeout.runtime_message = "engine #{@engine.name} ran past #{ENGINE_TIMEOUT} seconds and was killed"
```

Comment:
I'm pretty sure this approach won't work, but I have other ideas. I can flesh them out if we green-light this PR in general.